Add support for PermissionStatus.onchange

RupinMittal opened the pull request:
#### 7f531f3e0265f2d2722545e3600d39eb08242190
<pre>
Add support for PermissionStatus.onchange
<a href="https://bugs.webkit.org/show_bug.cgi?id=244633">https://bugs.webkit.org/show_bug.cgi?id=244633</a>

Reviewed by Chris Dumez and Youenn Fablet.

The PermissionStatus object returned by Permissions::query should
have an onchange event as specified by the Permissions API spec.
This patch adds this by adding a permissionChanged function to the
WKWebViewPrivate SPI which a client can call to indicate to WebKit
that a permission state has changed. WebPermissionController will
then query for the specified permission state and dispatch an
onchange event to the PermissionStatus object if the permission
state has indeed changed.

The permissionChanged function is a class/static function, so the
client needs to call it only once and WebKit will look through all
of the web processes and find the ones where the permission has
changed.

Currently, the onchange event is only supported in Window contexts.
A future patch will add support for Worker contexts.

An API test and Layout test were added to test this.

* LayoutTests/http/tests/notifications/permission/permission-status-onchange-event-expected.txt: Added.
* LayoutTests/http/tests/notifications/permission/permission-status-onchange-event.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/permissions/idlharness.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/permissions/idlharness.any.worker-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/permissions/PermissionController.h:
* Source/WebCore/Modules/permissions/PermissionObserver.h:
* Source/WebCore/Modules/permissions/PermissionStatus.h:
* Source/WebCore/Modules/permissions/PermissionStatus.idl:
* Source/WebCore/Modules/permissions/Permissions.cpp:
(WebCore::Permissions::sourceFromContext):
(WebCore::Permissions::toPermissionName):
(WebCore::Permissions::query):
(WebCore::sourceFromContext): Deleted.
* Source/WebCore/Modules/permissions/Permissions.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPagePermissionChanged):
* Source/WebKit/UIProcess/API/C/WKPagePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(+[WKWebView _permissionChanged:forOrigin:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/WebPermissionControllerProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::permissionChanged):
(WebKit::WebProcessProxy::sendPermissionChanged):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
(WebKit::WebPermissionController::create):
(WebKit::WebPermissionController::WebPermissionController):
(WebKit::WebPermissionController::~WebPermissionController):
(WebKit::WebPermissionController::query):
(WebKit::WebPermissionController::permissionChanged):
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.messages.in: Copied from Source/WebCore/Modules/permissions/PermissionStatus.idl.
* Source/WebKit/WebProcess/WebProcess.cpp:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/PermissionsAPI.mm:
(-[PermissionChangedTestAPIMessageHandler userContentController:didReceiveScriptMessage:]):
(-[PermissionChangedTestAPIUIDelegate _webView:queryPermission:forOrigin:completionHandler:]):
(TestWebKitAPI::TEST):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::grantNotificationPermission):
(WTR::TestController::denyNotificationPermission):

Canonical link: <a href="https://commits.webkit.org/254193@main">https://commits.webkit.org/254193@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d79eabdaa9187dfed9df7f8fdcde34a9f7a89ac3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/88376 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/32817 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/19211 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/97560 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/153033 "Found 1 new test failure: imported/blink/paint/deprecatedpaintlayer/non-self-painting-layer-overrides-visibility.html (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/92341 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/31313 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26963 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/80583 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/92218 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93983 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/24912 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/75225 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24882 "layout-tests (failure)") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/79812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/79928 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/67846 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/28926 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/13909 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/67/builds/28907 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/14932 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/2953 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/32107 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/75225 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/31059 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/34045 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -3928,6 +3933,37 @@ - (void)_setSystemAudioCaptureState:(WKSystemAudioCaptureState)state completionH
+
+- (void)_permissionMayHaveChanged:(NSString *)permissionName forOrigin:(WKSecurityOrigin *)origin
+{
+    auto name = [[WKWebView class] permissionNameFrom:permissionName];
```
Doesn't this work?
[WKWebView permissionNameFrom:permissionName];

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -3928,6 +3933,37 @@ - (void)_setSystemAudioCaptureState:(WKSystemAudioCaptureState)state completionH
+    auto topOrigin = origin->_securityOrigin->securityOrigin();
+
+    for (auto& webProcessPool : webProcessPools) {
+        for (auto& webProcessProxy : webProcessPool.get().processes())
```
webProcessPool->processes()

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -94,4 +102,24 @@ void WebPermissionController::permissionChanged(const WebCore::ClientOrigin& ori
+            auto source = WebCore::Permissions::sourceFromContext(context);
+            // FIXME: Add support for service workers and shared workers.
+            ASSERT(source && (*source == WebCore::PermissionQuerySource::Window || *source == WebCore::PermissionQuerySource::DedicatedWorker));
+            auto* page = downcast<WebCore::Document>(context).page();
```
How is this downcast<> correct in the dedicated worker case? I would think the context would be a DedicatedWorkerGlobalScope, not a Document.

Maybe you should drop the `|| *source == WebCore::PermissionQuerySource::DedicatedWorker` in your assertion above. I don't believe you support dedicated workers yet.

RupinMittal replied:
You're right, the downcast would need to be to DedicatedWorkerGlobal scope (I could do a switch case to get the page depending on the source). I think we do support dedicated workers though right? Because they have a page?

cdumez replied:
No, you have a main thread check in PermissionStatus when registering/unregistering the observer. For dedicated workers, this main thread check would fail and thus the worker's PermissionStatus currently don't register as observers.

RupinMittal replied:
Ah yes, that's correct. I'll make the change to indicate that this patch is only adding support for Window contexts.

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -94,4 +102,24 @@ void WebPermissionController::permissionChanged(const WebCore::ClientOrigin& ori
+        if (observer.origin().topOrigin == topOrigin && observer.descriptor().name == permissionName) {
+            auto& context = observer.context();
+            auto source = WebCore::Permissions::sourceFromContext(context);
+            // FIXME: Add support for service workers and shared workers.
```
How about dedicated workers? I don't believe we support any workers yet?

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -94,4 +102,24 @@ void WebPermissionController::permissionChanged(const WebCore::ClientOrigin& ori
+            auto* page = downcast<WebCore::Document>(context).page();
+
+            if (page) {
+                query(WebCore::ClientOrigin { topOrigin, topOrigin }, WebCore::PermissionDescriptor { permissionName }, page, *source, [&observer](auto newState) {
```
I don't believe using topOrigin twice is correct here. You need to use the correct frame origin for the second member, which would likely be `downcast<WebCore::Document>(context)->securityOrigin()`

cdumez replied:
Also, you cannot capture observer by reference in an async lambda.

szewai commented on Source/WebCore/Modules/permissions/PermissionController.h:
```diff
@@ -40,13 +40,14 @@ struct PermissionDescriptor;
     virtual void query(WebCore::ClientOrigin&&, PermissionDescriptor&&, Page*, PermissionQuerySource, CompletionHandler<void(std::optional<PermissionState>)>&&) = 0;
     virtual void addObserver(PermissionObserver&) = 0;
     virtual void removeObserver(PermissionObserver&) = 0;
+    virtual bool isWebPermissionController() const = 0;
```
You may return false by default here and only return true in WebPermissionController::isWebPermissionController().

szewai commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -3928,6 +3933,37 @@ - (void)_setSystemAudioCaptureState:(WKSystemAudioCaptureState)state completionH
+    return std::nullopt;
+}
+
+- (void)_permissionMayHaveChanged:(NSString *)permissionName forOrigin:(WKSecurityOrigin *)origin
```
The SPI should be called permissionChanged in my opinion because client calls it when some permission has changed; for each PermissionController it can be permissionMayHaveChanged.

szewai commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -3928,6 +3933,37 @@ - (void)_setSystemAudioCaptureState:(WKSystemAudioCaptureState)state completionH
     });
 }
 
++ (std::optional<WebCore::PermissionName>)permissionNameFrom:(NSString *)permissionName
```
`static std::optional<WebCore::PermissionName> toPermissionName(NSString* name)`

szewai commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h:
```diff
@@ -44,13 +45,16 @@ class WebPermissionController final : public CanMakeWeakPtr<WebPermissionControl
     void query(WebCore::ClientOrigin&&, WebCore::PermissionDescriptor&&, WebCore::Page*, WebCore::PermissionQuerySource, CompletionHandler<void(std::optional<WebCore::PermissionState>)>&&) final;
     void addObserver(WebCore::PermissionObserver&) final;
     void removeObserver(WebCore::PermissionObserver&) final;
+    bool isWebPermissionController() const final;
```
You can return true here

szewai commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -3928,6 +3933,37 @@ - (void)_setSystemAudioCaptureState:(WKSystemAudioCaptureState)state completionH
+
+    for (auto& webProcessPool : webProcessPools) {
+        for (auto& webProcessProxy : webProcessPool.get().processes())
+            webProcessProxy->webPermissionController()->permissionMayHaveChanged(*name, topOrigin);
```
Why calling `permissionMayHaveChanged` on `WebPermissionControllerProxy`?
All WebPermissionControllerProxy:: permissionMayHaveChanged does is calling WebProcessProxy:: permissionMayHaveChanged; so you can just call `webProcessProxy->permissionMayHaveChanged()` here

szewai commented on Source/WebKit/UIProcess/WebProcessProxy.h:
```diff
@@ -451,6 +452,10 @@ class WebProcessProxy : public AuxiliaryProcessProxy, private ProcessThrottlerCl
 
+    std::unique_ptr<WebPermissionControllerProxy>& webPermissionController() { return m_webPermissionController; }
+
+    void permissionMayHaveChanged(WebCore::PermissionName, const WebCore::SecurityOriginData&);
```
No need for these.

szewai commented on Source/WebKit/WebProcess/WebProcess.cpp:
```diff
@@ -2166,6 +2166,12 @@ RemoteMediaEngineConfigurationFactory& WebProcess::mediaEngineConfigurationFacto
+void WebProcess::permissionMayHaveChanged(WebCore::PermissionName permissionName, const WebCore::SecurityOriginData& topOrigin)
+{
+    auto& webPermissionController = downcast<WebPermissionController>(WebCore::PermissionController::shared());
+    webPermissionController.permissionMayHaveChanged(permissionName, topOrigin);
```
I think `permissionMayHaveChanged` should be virtual function on `WebCore::PermissionController` like `addObserver`/`removeObserver`

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/PermissionsAPI.mm:
```diff
@@ -102,4 +124,44 @@ static void urlEncodeIfNeeded(uint8_t byte, StringBuilder& buffer)
+    auto delegate = adoptNS([[PermissionChangedTestAPIUIDelegate  alloc] init]);
+    [webView setUIDelegate:delegate.get()];
+
+    NSString *script = @"navigator.permissions.query({ name : 'notifications' }).then((permissionStatus) => {"
```
Please use other permission (e.g. geolocation) as notifications is only enabled on macOS.

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/PermissionsAPI.mm:
```diff
@@ -102,4 +124,44 @@ static void urlEncodeIfNeeded(uint8_t byte, StringBuilder& buffer)
 
+TEST(PermissionsAPI, PermissionMayHaveChanged)
+{
+    [TestProtocol registerWithScheme:@"https"];
```
Is this needed?

RupinMittal replied:
The way I had written the request in this iteration of the patch--this line was needed. But there's a better way to write the request that doesn't require this. I'll use that better version in the next iteration of this patch.

szewai commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/PermissionsAPI.mm:
```diff
@@ -102,4 +124,44 @@ static void urlEncodeIfNeeded(uint8_t byte, StringBuilder& buffer)
+
+    [webView performAfterReceivingMessage:@"prompt" action:^{
+        wasPermissionStatusChangedToPrompt = true;
+    }];
```
I would recommend using EXPECT to check if state (message) is expected instead of waiting until these values are true. Timeout is less preferred than failure cause it's harder to debug and it delays the run of other tests. (With current impl, the test can time out if the page sends back "FAIL" message, or if state is unexpected, or if page never sends back message).

cdumez commented on Source/WebCore/Modules/permissions/PermissionStatus.h:
```diff
@@ -56,9 +56,11 @@ class PermissionStatus final : public PermissionObserver, public ActiveDOMObject
     void stateChanged(PermissionState) final;
     const ClientOrigin& origin() const final { return m_origin; }
     const PermissionDescriptor& descriptor() const final { return m_descriptor; }
+    const ScriptExecutionContext& context() const final { return *ActiveDOMObject::scriptExecutionContext(); }
```
It is not safe here to dereference ActiveDOMObject::scriptExecutionContext() unconditionally. This function should probably return a raw pointer, not a reference.

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -86,11 +89,23 @@ void WebPermissionController::removeObserver(WebCore::PermissionObserver& observ
     for (auto& observer : m_observers) {
-        if (observer.origin() == origin && observer.descriptor() == descriptor)
-            observer.stateChanged(state);
+        if (observer.origin().topOrigin == topOrigin && observer.descriptor().name == permissionName) {
```
Can we reverse the checks? The `observer.descriptor().name == permissionName` check is cheaper so it makes more sense to try this one first.

cdumez replied:
Also, in WebKit we like early returns to reduce scoping.

This should be better as:
```
if (observer.descriptor().name != permissionName || observer.origin().topOrigin != topOrigin)
    continue;
```

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -86,11 +89,23 @@ void WebPermissionController::removeObserver(WebCore::PermissionObserver& observ
-        if (observer.origin() == origin && observer.descriptor() == descriptor)
-            observer.stateChanged(state);
+        if (observer.origin().topOrigin == topOrigin && observer.descriptor().name == permissionName) {
+            auto& context = observer.context();
```
context can totally be null here.

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -86,11 +89,23 @@ void WebPermissionController::removeObserver(WebCore::PermissionObserver& observ
+            ASSERT(source && *source == WebCore::PermissionQuerySource::Window);
+            auto* page = downcast<WebCore::Document>(context).page();
+
+            if (page) {
```
```
if (!page)
    continue;
```

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -86,11 +89,23 @@ void WebPermissionController::removeObserver(WebCore::PermissionObserver& observ
+            auto* page = downcast<WebCore::Document>(context).page();
+
+            if (page) {
+                query(WebCore::ClientOrigin { observer.origin() }, WebCore::PermissionDescriptor { permissionName }, page, *source, [observer = WeakPtr { observer }](auto newState) {
```
PermissionDescriptor is just an enum, no need to pass it as r-value reference. Please update query() to take a `PermissionDescriptor` instead of a `PermissionDescriptor&&`. Then you can drop the `WebCore::PermissionDescriptor { }` here.

RupinMittal replied:
That's true. PermissionDescriptor is a struct that only contains a PermissionName, so there's no need to pass it in as an r-value reference. But I believe we can't drop the `WebCore::PermissionDescriptor { }`.

cdumez commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
```diff
@@ -86,11 +89,23 @@ void WebPermissionController::removeObserver(WebCore::PermissionObserver& observ
+
+            if (page) {
+                query(WebCore::ClientOrigin { observer.origin() }, WebCore::PermissionDescriptor { permissionName }, page, *source, [observer = WeakPtr { observer }](auto newState) {
+                    if (newState && *newState != observer->currentState())
```
You need to check that observer is non-null here since it is a WeakPtr.

Also, the newState check is useless here. The operator==() for std::optional already does this for you. This should be:
```
if (observer && newState != observer->currentState())
```

cdumez commented on Source/WebKit/WebProcess/WebProcess.messages.in:
```diff
@@ -209,4 +209,6 @@ messages -> WebProcess LegacyReceiver NotRefCounted {
     OpenDirectoryCacheInvalidated(WebKit::SandboxExtension::Handle handle)
 #endif
+
+    permissionChanged(enum:uint8_t WebCore::PermissionName permissionName, struct WebCore::SecurityOriginData topOrigin)
```
Ideally, this IPC would be on WebPermissionController, not on WebProcess. We have a dedicated class for this so we shouldn't pollute the generic WebProcess. You likely want to introduce a WebPermissionController.messages.in instead.

You can use WebLockRegistryProxy as reference.

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/PermissionsAPI.mm:
```diff
@@ -102,4 +131,41 @@ static void urlEncodeIfNeeded(uint8_t byte, StringBuilder& buffer)
     EXPECT_FALSE(didReceiveQueryPermission);
 }
 
+TEST(PermissionsAPI, PermissionMayHaveChanged)
```
In general, we try to avoid API tests unless they are strictly necessary. Why aren't we using layout tests here?
Seems like the existing testRunner.grantWebNotificationPermission() / denyWebNotificationPermission() could cause permissionChanged() to be called, then we should be able to use layout tests.

The reason we prefer API tests is that they are faster to run (can be parallelized) and can be more easily skipped.

szewai replied:
I think API test is used to ensure WebKit API works correctly; layout test is used to ensure Web API work correctly.
testRunner.grantWebNotificationPermission() is used to simulate client's behavior in a layout test. We are not testing API in layout test; we are assuming it works. And we won't be able to check how many times queryPermission is called in layout test.

We should probably adopt permissionChanged in grantWebNotificationPermission/denyWebNotificationPermission though, to make TestRunner act as a functional client.

cdumez replied:
> we won't be able to check how many times queryPermission is called in layout test
Does this API test really check that?

> I think API test is used to ensure WebKit API works correctly; layout test is used to ensure Web API work correctly.
Here the API is really here to support a Web feature though: The firing of a change event.

If you want an API test to validate that the API behaves correctly, and that queryPermission is only called once, fine. However, this needs layout testing too then.
Future tests involving workers, dedicated workers and service workers should definitely be layout tests and not API tests.

cdumez replied:
The change event is a Web feature, not a native API feature. Ideally we'd be using WPT tests with a WebDriver hook to change the permissions.

szewai replied:
Yeah the patch has both WebKit API and Web API changes, so we should probably add both tests

lauromoura commented on Source/WebKit/DerivedSources.make:
```diff
@@ -225,6 +225,7 @@ MESSAGE_RECEIVERS = \
 	WebProcess/WebCoreSupport/WebBroadcastChannelRegistry \
 	WebProcess/WebCoreSupport/WebDeviceOrientationUpdateProvider \
 	WebProcess/WebCoreSupport/WebFileSystemStorageConnection \
+	WebProcess/WebCoreSupport/WebPermissionController \
```
An entry like this should be added in the `Source/WebKit/CMakeLists.txt`, in the `WebCoreSupport` block around lines 321-324 to trigger the Messages derived sources generation for `WebPermissionController`, no?

The PR as it is is failing in both GTK/WPE, leaving the build in a inconsistent state: https://ews-build.webkit.org/#/builders/36/builds/80200

RupinMittal replied:
Yes, you're right. It looks like I added this entry in `WebKit/DerivedSources.make` but forgot to add it to `WebKit/CMakeLists.txt`. I'll add it there as well. Thank you!

youennf commented on Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h:
```diff
@@ -71,3 +80,7 @@ class WebPermissionController final : public CanMakeWeakPtr<WebPermissionControl
 
 } // namespace WebCore
+
+SPECIALIZE_TYPE_TRAITS_BEGIN(WebKit::WebPermissionController)
```
Is this needed, I do not see a downcast from PermissionController?

RupinMittal replied:
Nope, it was needed in a previous iteration of this patch, but not anymore. Thank you for the catch!

youennf reviewed (approved):
LGTM

cdumez reviewed (approved):
LGTM

RupinMittal committed:
Add support for PermissionStatus.onchange
https://bugs.webkit.org/show_bug.cgi?id=244633

Reviewed by Chris Dumez and Youenn Fablet.

The PermissionStatus object returned by Permissions::query should
have an onchange event as specified by the Permissions API spec.
This patch adds this by adding a permissionChanged function to the
WKWebViewPrivate SPI which a client can call to indicate to WebKit
that a permission state has changed. WebPermissionController will
then query for the specified permission state and dispatch an
onchange event to the PermissionStatus object if the permission
state has indeed changed.

The permissionChanged function is a class/static function, so the
client needs to call it only once and WebKit will look through all
of the web processes and find the ones where the permission has
changed.

Currently, the onchange event is only supported in Window contexts.
A future patch will add support for Worker contexts.

An API test and Layout test were added to test this.

* LayoutTests/http/tests/notifications/permission/permission-status-onchange-event-expected.txt: Added.
* LayoutTests/http/tests/notifications/permission/permission-status-onchange-event.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/permissions/idlharness.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/permissions/idlharness.any.worker-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/permissions/PermissionController.h:
* Source/WebCore/Modules/permissions/PermissionObserver.h:
* Source/WebCore/Modules/permissions/PermissionStatus.h:
* Source/WebCore/Modules/permissions/PermissionStatus.idl:
* Source/WebCore/Modules/permissions/Permissions.cpp:
(WebCore::Permissions::sourceFromContext):
(WebCore::Permissions::toPermissionName):
(WebCore::Permissions::query):
(WebCore::sourceFromContext): Deleted.
* Source/WebCore/Modules/permissions/Permissions.h:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPagePermissionChanged):
* Source/WebKit/UIProcess/API/C/WKPagePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(+[WKWebView _permissionChanged:forOrigin:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/WebPermissionControllerProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.cpp:
(WebKit::WebProcessProxy::permissionChanged):
(WebKit::WebProcessProxy::sendPermissionChanged):
* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp:
(WebKit::WebPermissionController::create):
(WebKit::WebPermissionController::WebPermissionController):
(WebKit::WebPermissionController::~WebPermissionController):
(WebKit::WebPermissionController::query):
(WebKit::WebPermissionController::permissionChanged):
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.h:
* Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.messages.in: Copied from Source/WebCore/Modules/permissions/PermissionStatus.idl.
* Source/WebKit/WebProcess/WebProcess.cpp:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/PermissionsAPI.mm:
(-[PermissionChangedTestAPIMessageHandler userContentController:didReceiveScriptMessage:]):
(-[PermissionChangedTestAPIUIDelegate _webView:queryPermission:forOrigin:completionHandler:]):
(TestWebKitAPI::TEST):
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::grantNotificationPermission):
(WTR::TestController::denyNotificationPermission):

Canonical link: https://commits.webkit.org/254193@main

geoffreygaren commented on Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
```diff
@@ -3938,6 +3941,15 @@ - (CGFloat)_overrideDeviceScaleFactor
     return _page->customDeviceScaleFactor().value_or(0);
 }
 
++ (void)_permissionChanged:(NSString *)permissionName forOrigin:(WKSecurityOrigin *)origin
+{
+    auto name = WebCore::Permissions::toPermissionName(permissionName);
+    if (!name)
+        return;
+
+    WebKit::WebProcessProxy::permissionChanged(*name, origin->_securityOrigin->securityOrigin());
+}
+
```
I see a problem with this design: It's process-global. 

In our software, a process is typically composed of code written by many different people and teams. There's application code (and the application may be split into many parts), and also code from different frameworks.

This means that, within a given process, there may be many WebProcesses serving many different purposes. A person invoking a process-global API may trigger effects on WebViews that they didn't expect and weren't even aware of. As a result, process-global API almost always causes some sort of problem in the end. 

At a minimum, this design is inefficient. For example, it will send IPC to the Foundation NSAttributed string WebProcess, to cached, back-forward cached, pre-loaded, and pre-warmed web processes, to extensions WebProcesses, and so on when Safari changes permissions -- for no reason. In Chrome on iOS, a change in an Incognito WebProcess's settings will send IPC a standard WebProcess, even though they don't share settings.

At some point, someone is likely to be surprised by the effects of one of those messages, and assert or crash as a result.

I think we should reformulate this API to be per-WebView, and have the client app invoke the API on the WebViews it wishes to include.

szewai replied:
> At a minimum, this design is inefficient.

True, we might optimize it if UI process only sends message to web processes that has queried permission or has permission observers. 

> At some point, someone is likely to be surprised by the effects of one of those messages, and assert or crash as a result.

Not sure why this would happen. What are they trying to assert or check? I think the messages are handled by WebKit. 

> I think we should reformulate this API to be per-WebView, and have the client app invoke the API on the WebViews it wishes to include.

The reason we don't do this is client cannot notify SharedWorker and ServiceWorker about the change. 
For permission in these workers, we pick the most "reasonable" WebView to query permission. If we take WebView as parameter, when the permission changes for the WebView, we don't know if we should update state in the workers.

geoffreygaren replied:
> True, we might optimize it if UI process only sends message to web processes that has queried permission or has permission observers.

This is still inefficient in a process where multiple WebViews from distinct frameworks all make use of permissions. To the extent that it might wake up a process (or 100!) that was otherwise sleeping, the inefficiency may sometimes be significant.

> > At some point, someone is likely to be surprised by the effects of one of those messages, and assert or crash as a result.
> Not sure why this would happen. What are they trying to assert or check? I think the messages are handled by WebKit.

This behavior is observable to a client because it will sometimes call out to the client to ask for permission status. Right?

There are just so many ways this could go wrong. 

A simple example is that a WebKitTestRunner that attempts to print out delegate callbacks for unit testing might randomly succeed or fail based on extra delegate callbacks triggered by a totally unrelated test case.

> > I think we should reformulate this API to be per-WebView, and have the client app invoke the API on the WebViews it wishes to include.
> The reason we don't do this is client cannot notify SharedWorker and ServiceWorker about the change.
> For permission in these workers, we pick the most "reasonable" WebView to query permission. If we take WebView as parameter, when the permission changes for the WebView, we don't know if we should update state in the workers.

Can WebKit infer which workers to notify based on which WebViews were notified?
